types/tx.go

```diff
@@ -44,9 +44,17 @@ type Txs []Tx
 
 // Hash returns the Merkle root hash of the transaction hashes.
 // i.e. the leaves of the tree are the hashes of the txs.
+// TODO(@raneet10): Investigate and decide whether we should handle empty or single tx specially
 func (txs Txs) Hash() []byte {
-	hl := txs.hashList()
-	return merkle.HashFromByteSlices(hl)
+	switch len(txs) {
+	case 0:
+		return nil
+	case 1:
+		return txs[0].Hash()
+	default:
+		hl := txs.hashList()
+		return merkle.HashFromByteSlices(hl)
+	}
```

Comment:
Can it be kept in synch with upstream?

Reply:
If we hard-fork, it would be a golden opportunity to remove this deviation from vanilla CometBFT. As we discussed some week ago, if we keep this different from vanilla CometBFT, then it will be problematic to come back to it, forcing us to keep this changed forever (or until the next hard-fork, which will hopefully never happen again)

Reply:
Yeah my thoughts exactly!

Reply:
Resolved